lib/remote/apilistener-filesync.cpp

```diff
@@ -667,35 +663,81 @@ bool ApiListener::CheckConfigChange(const ConfigDirInformation& oldConfig, const
 		<< "' vs. new (" << newChecksums->GetLength() << "): '"
 		<< JsonEncode(newChecksums) << "'.";
 
-	// Don't check for different length, this may be influenced from internal files
-	ObjectLock olock(oldChecksums);
 
-	for (const Dictionary::Pair& kv : oldChecksums) {
-		String path = kv.first;
-		String oldChecksum = kv.second;
+	/* Since internal files are synced here too, we can not depend on length.
+	 * So we need to go through both checksum sets to cover the cases"everything is new" and "everything was deleted".
+	 */
+	{
+		ObjectLock olock(oldChecksums);
+		for (const Dictionary::Pair& kv : oldChecksums) {
+			String path = kv.first;
+			String oldChecksum = kv.second;
 
-		// TODO: Figure out if config changes only apply to '.conf'. Leaving this open for other config files.
-		//if (!Utility::Match("*.conf", path))
-		//	continue;
+			// TODO: Figure out if config changes only apply to '.conf'. Leaving this open for other config files.
+			//if (!Utility::Match("*.conf", path))
+			//	continue;
 
-		/* Ignore internal files, especially .timestamp and .checksums.
-		 *
-		 * If we don't, this results in "always change" restart loops.
-		 */
-		if (Utility::Match("/.*", path))
-			continue;
+			/* Ignore internal files, especially .timestamp and .checksums.
+			 *
+			 * If we don't, this results in "always change" restart loops.
+			 */
+			if (Utility::Match("/.*", path)) {
+				Log(LogDebug, "ApiListener")
+						<< "Ignoring internal file '" << path << "'";
+				continue;
+			}
 
-		Log(LogDebug, "ApiListener")
-			<< "Checking " << path << " for checksum: " << oldChecksum;
+			Log(LogDebug, "ApiListener")
+					<< "Checking " << path << " for old checksum: " << oldChecksum;
+
+			// Check if key exists first for more verbose logging.
+			// TODO: Don't do this later on.
+			if (!newChecksums->Contains(path)) {
+				Log(LogDebug, "ApiListener")
+						<< "File '" << path << "' was deleted by remote.";
+				return true;
+			}
 
-		// Check whether our key exists in the new checksums, and they have an equal value.
-		String newChecksum = newChecksums->Get(path);
+			String newChecksum = newChecksums->Get(path);
+			if (newChecksum != kv.second) {
+				Log(LogDebug, "ApiListener")
+						<< "Path '" << path << "' doesn't match old checksum '"
+						<< oldChecksum << "' with new checksum '" << newChecksum << "'.";
+				return true;
+			}
+		}
+	}
+
+	{
+		ObjectLock olock(newChecksums);
+		for (const Dictionary::Pair& kv : newChecksums) {
```

Comment:
This duplicates the code, imho it is sufficient to pick the Dictionary with the greater length and iterate over it. It then can check against matches in the other Dictionary.

See https://github.com/Icinga/icinga2/pull/7306/commits/6af65a72488c8c136ad90d79c594c0c1a18d9d03 how you would assign the "generic" Dictionary for the loop depending on the size, and then iterate over it.

Reply:
The issue is that newChecksums has two additional checksums, so the following would lead to "no changed config" when not having the second loop and only going with the larger dictionary:
oldChecksums: {a, b, c, d}
newChecksums: {.timestamp, .checksums, a, b, c }

Reply:
Ok, then please move this at least into a generic function which does the loop and checks. This avoids code duplication and possible errors.

Reply:
I thought about that, the two loops are different. The first one compares checksums but the second does not care about that. So a function comareChecksums() would only need to be called once. As in the whole block becomes a function.

Reply:
Ah, ok. After re-reading the code again, it makes sense.